pyomo/contrib/trustregion/TRF.py

```diff
@@ -122,128 +102,312 @@ def TRF(m, eflist, config):
+    ))
+
+    return CONFIG
+
+
+@SolverFactory.register(
+    'trustregion',
+    doc='Trust region algorithm "solver" for black box/glass box optimization')
+class TrustRegionSolver(object):
+    """
+    The Trust Region Solver is a 'solver' based on the 2016/2018/2020 AiChE
+    papers by Eason (2016/2018), Yoshio (2020), and Biegler.
 
-                config.trust_radius = min(
-                    max(config.gamma_e*stepNorm, config.trust_radius),
-                    config.radius_max)
+    """
 
-            else:
-                logger.iterlog.thetaStep = True
-
-                fe = FilterElement(
-                    obj - config.gamma_f*theta,
-                    (1 - config.gamma_theta)*theta)
-                filteR.addToFilter(fe)
-
-                # Calculate rho for theta step trust region update
-                rhok = 1 - ((theta - config.ep_i) /
-                            max(thetak, config.ep_i))
-                if rhok < config.eta1:
-                    config.trust_radius = max(config.gamma_c*stepNorm,
-                                      config.delta_min)
-                elif rhok >= config.eta2:
-                    config.trust_radius = min(
-                        max(config.gamma_e*stepNorm, config.trust_radius),
-                        config.radius_max)
-
-        # Accept step
-        rebuildROM = True
-        xk, yk, zk = cloneXYZ(x, y, z)
-        thetak = theta
-        objk = obj
-
-
-    logger.printVectors()
-#    problem.reverseTransform()
+    def __init__(self, **kwds):
+        self._CONFIG = _trf_config()
+        self._CONFIG.set_value(kwds)
+
+    def available(self, exception_flag=True):
+        """
+        Check if solver is available.
+        """
+        return True
+
+    def version(self):
+        """
+        Return a 3-tuple describing the solver version.
+        """
+        return __version__
+
+    def license_is_valid(self):
+        """
+        License for using Trust Region solver.
+        """
+        return True
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, et, ev, tb):
+        pass
+
+    def solve(self, model, degrees_of_freedom_variables,
+              ext_fcn_surrogate_map_rule=None, **kwds):
+        """
+        This method calls the TRF algorithm.
+
+        Parameters
+        ----------
+        degrees_of_freedom_variables : List of Vars
+            User-supplied input. The user must provide a list of vars which
+            are the degrees of freedom or decision variables within
+            the model.
+        ext_fcn_surrogate_map_rule : Function, optional
+            In the 2020 Yoshio/Biegler paper, this is refered to as
+            the basis function `b(w)`.
+            This is the low-fidelity model with which to solve the original
+            process model problem and which is integrated into the
+            surrogate model.
+            The default is 0 (i.e., no basis function rule.)
+
+        """
```

Comment:
We should consider adding the CONFIG documentation to this docstring using `add_docstring_list`

Reply:
Done in my `trust_region_docs` branch.